### DISCUSSION TITLE
Keyboard shortcut documentation for editor in the Wagtail UI

### DISCUSSION THREAD

**Discussion by tmsndrs** [wagtail/wagtail#12050] [Ideas]
### Is your proposal related to a problem?

As part of the ATAG guidelines.

(Guideline A.3.1) (For the authoring tool user interface) [Provide keyboard access to authoring features](https://www.w3.org/TR/ATAG20/#gl_a31). 

We want to work towards a fully compliant, customisable and flexible keyboard shortcut system within Wagtail.

At this time, there are some undocumented keyboard shortcuts, some browser 'default' shortcuts and lots of available shortcuts in Draftail.

These, however, are hard to discover and we would like to improve this.

### Describe the solution you'd like

As per the comments below, here are some mockups for how a shortcuts panel could work.

Access from the Help menu (help users to learn the command to toggle the panel to negate the need to access it from this menu):
<img width="815" alt="image" src="https://github.com/wagtail/wagtail/assets/4907819/f4356a6a-c623-4f9e-ad52-11322e2caa9f">

Panel: 
- Scrollable
- Considered making it searchable, but decided to keep things simple for now
- Link out to the full documentation
- Prioritise Wagtail specific commands (currently only 2, bit would be great to add more here eg: Duplicate block, Move block up/down, Add block (outside of richtext), Collapse/Expand block/field, Collapse/Expand all, Toggle comments panel, Toggle status panel, Toggle Minimap ...
- [Prototype here](https://www.figma.com/proto/h67EsVXdWsfu38WGGxWfpi/Wagtail-CMS?type=design&node-id=14420-153315&viewport=6289%2C-12853%2C1&t=Sk9k7iALDZCA1bu7-0&scaling=scale-down-width&starting-point-node-id=14420%3A153315&show-proto-sidebar=1&hide-ui=1) (use ⌘ + /) to open the panel in Figma

<img width="1117" alt="image" src="https://github.com/wagtail/wagtail/assets/4907819/0738df3e-ebfb-4124-9001-525aeba49977">

#### Tasks

* [x] Adopt a Stimulus controller approach to add keyboard shortcuts to specific elements - https://github.com/wagtail/wagtail/issues/10167
* [ ] Enhance the Stimulus controller to allow for a tooltip on hover to show the relevant keyboard shortcut on elements that have it.
* [x] Ensure that there is a way to allow a keyboard shortcut to trigger globally, even in text input, via a Stimulus attribute. https://github.com/wagtail/wagtail/issues/11844
* [ ] Migrate any other existing (non-Draftail) keyboard shortcuts to the Stimulus approach, or some other mechanism so a central registry on the client can be used if needed in the future.
* [ ] Add a keyboard shortcut for the sidebar opening/closing (likely `[`/`]`)
* [ ] Add a keyboard shortcut for the minimap, when available, opening/closing + some other side panel keyboard shortcuts, global open/closing of panels. More in the solution above.
* [x] Add a very basic, starting point, keyboard shortcut overview dialog https://github.com/wagtail/wagtail/issues/11711
* [x] Enhance the keyboard shortcuts dialog with a more suitable 'keyboard' icon #11767
* [ ] Add a way to trigger the keyboard shortcut dialog with a keyboard shortcut (either `cmd + /` or `shift + /` or both) & include on the dialog itself
* [ ] Once the editor's guide has the keyboard shortcuts documented on a page, link to that page in the dialog - see https://github.com/wagtail/guide/issues/368 & https://github.com/wagtail/guide/issues/346
* [ ] Develop a way to ensure that the keyboard shortcut (groups?) only show on admin pages where they are relevant (needs discussion on implementation, could be a simple template block approach or a more complex hooks or JS approach), once built, we should ensure that the existing admin code adopts this and also document.
* [ ] Make the content within the keyboard shortcut dialog searchable with a simple text field and filtering/highlighting, this can be done with an existing Stimulus controller or something new if needed.
* [ ] Long term: Add support for 'activating' the action that the keyboard shortcut has. E.g. you could activate the minimap by opening the keyboard shortcuts dialog and clicking the minimap action.  At this point we may want to raise a new issue as this is more of a 'slash commands' for global actions feature and will need more thought.

### Describe alternatives you've considered

- This seems to be a long term direction and we would like to see this in place, in some form, in a future release in Wagtail.

### Additional context

- Many fixes to keyboard shortcuts within the rich text editor update were fixed (issues documented #2191 ). 
- Very early mockup from the original issue below.

<img width="812" alt="screen shot 2017-10-19 at 17 24 29" src="https://user-images.githubusercontent.com/7290588/31782427-c51ff1ac-b4f2-11e7-9e2b-a21b06f51f1c.png">

### Working on this

- For now, we will split out smaller issues for others to pick up either within the core team or other contributors.

**Comment by tmsndrs** [comment]
Note: I haven't thought about how this modal would be launched yet because I'm unclear if we'll have one set of shortcuts (preferable from a UX perspective I think) or different shortcuts within different parts of the UI.

**Comment by lb-** [comment]
Just adding some basic research into what other popular web applications do in regards activation of keyboard shortcuts modal (via a keyboard shortcut).

It seems that Medium's modal is the only one that can actually be activated within a text box that the user has focus on.

**Medium**
* Open Shortcuts Modal `⌘ + ?` / `Ctrl + ?`
* Close Shortcuts Modal `⌘ + ?` / `Ctrl + ?` not Escape
* Works if you are in the middle of writing text or not.
* Note: Great interface, simple and lets you slide to more advanced shortcuts and small animations on how to do basic things like format text.
* https://help.medium.com/hc/en-us/articles/214672207-Keyboard-shortcuts

**Inbox by Google, Google Drive, Gmail**
* Open Shortcuts Modal `Shift + ?`
* Close Shortcuts Modal `Esc`
* Disabled when inside a text box.
* https://support.google.com/mail/answer/6594?hl=en-GB

**Trello**
* Open Shortcuts Modal `Shift + ?`
* Close Shortcuts Modal `Esc`
* Disabled when writing inside a text box.
* http://help.trello.com/article/798-using-keyboard-shortcuts-in-trello

**Github**
* Open Shortcuts Modal `Shift + ?`
* Close Shortcuts Modal `Esc`
* Disabled when writing inside a text box.
* https://help.github.com/articles/using-keyboard-shortcuts/

**Comment by thibaudcolas** [comment]
Here is the list of keyboard shortcuts supported by Draftail, heavily inspired by Google Docs: https://github.com/springload/draftail/tree/master/docs#keyboard-shortcuts.

I foresee problems integrating this with the rich text editor because the shortcuts would need to change depending on:

- Which editor is used on the current page
- Which rich text features are enabled on the current editor(s)

I imagine we could ~easily~ circumvent this though by having some way for rich text editors to hook into the "show keyboard shortcuts" modal, and insert the relevant shortcuts when people are triggering this from a given rich text editor.

**Comment by thibaudcolas** [comment]
It's easier to see how this could work with the improvements done over the last few months:

- The shortcut modal opens with a keyboard shortcut like in examples above
- By default it display's "Wagtail's" shortcuts like in the original screenshot

For rich text shortcuts,

- The shortcut screen hooks into the new "rich text features" API.
- Editors implementing a given rich text feature can define the corresponding shortcut.
- When the shortcut modal opens, if there are rich text features active on the page it grabs these and displays them in a separate section for each active editor.

---

I think we could start with only the "default Wagtail shortcuts" modal, then expand to rich text shortcuts later on. If someone is interested in implementing this with React, I'm more than happy to give plenty of guidance.

**Comment by SalahAdDin** [comment]
@thibaudcolas More documentation for `Wagtail 2.0`

**Comment by thibaudcolas** [comment]
@SalahAdDin there's now a list of all Draftail keyboard shortcuts over at https://www.draftail.org/docs/keyboard-shortcuts (which is referenced in the official Wagtail docs).

**Comment by phildexter** [comment]
@thibaudcolas 

Could we have this on hover?

Miro does this nicely. When you hover over an icon it shows you the shortcut.

![image](https://user-images.githubusercontent.com/34743669/81953890-1d105f00-9600-11ea-892f-66d82076563c.png)

![image](https://user-images.githubusercontent.com/34743669/81954374-ae7fd100-9600-11ea-950b-4ebfa3b72991.png)

Airtable also do this nicely
![image](https://user-images.githubusercontent.com/34743669/81971160-8f8d3900-9618-11ea-98ab-136e52880732.png)


I think we could do a similar thing in Wagtail. When you hover over the Save or Preview buttons we show the shortcut then? This would help with discoverability of these shortcuts too and could be rolled out elsewhere if we brought more in.

**Comment by thibaudcolas** [comment]
I like the idea! Especially considering it’s what Draftail already does, at least for shortcuts that are associated with buttons.

![shortcuts-tooltip](https://user-images.githubusercontent.com/877585/82016908-dcedc280-9679-11ea-84cf-0bd3a411b658.gif)

Do you think it needs further design or would we just replicate the pattern from Draftail elsewhere?

**Comment by phildexter** [comment]
Lovely, thanks @thibaudcolas !

@benenright what do you think? Are you happy with this approach? If so shall I get some designs done that we can pass by you, or maybe you could do the designs?

**Comment by allcaps** [comment]
This is the way Slack presents keyboard shortcuts. I like it.

<img width="1232" alt="Screenshot 2020-06-06 at 18 46 43" src="https://user-images.githubusercontent.com/1969342/83952050-4eaecb80-a836-11ea-9540-f4cf59684244.png">

Slack has some nice details:
- Main heading in combination with the toggle shortcut. _Press Cmd/ to toggle this panel._
- Shortcuts grouped per section. _Basics, Navigation, et cetera_
- The description is sometimes a link to perform the same action.
- There is an `?` icon for additional information.

Some keyboard shortcuts do _not_ have a visible element:
- Scroll up/down
- Move focus to next/previous section
- Undo/redo
- ...

Wagtail can have those as well. They can _only_ be communicated by a 'cheat sheet' as there is no element to hover.

I think a cheat sheet is the best solution because:
- An overview of all shortcuts is the only way to learn all shortcuts
- A centralised place to display all shortcuts is the best way to hook in custom or third party shortcuts.
- Supports shortcuts that do not have a visible element.

I'd like to help and make the cheat sheet happen. Especially the part where integrators and third parties can hook their own shortcuts to the cheat sheet.

**Comment by phildexter** [comment]
That's great feedback @allcaps. Especially the part about allowing third parties to be able to hook in too. I think a cheat sheet would be useful, but perhaps we could even do both (show some in context where it's useful, but all in the cheat sheet).

The more interesting issue for me is the inconsistency with which the existing shortcuts work for save and print. Cmd+P often opens the print dialogue, and Cmd+S tries to save the page. I realise now that this is because of where I'm focusing - if I click on the top of the page near the title and use the shortcuts it is fine. If I am editing a field and then try then it doesn't work.

This isn't great, as the majority of the time you'll want to save or preview once you've just made an edit. I think we shouldn't promote these shortcuts until we figure out a way to make this more consistent, otherwise we risk detracting from Wagtail overall (Kano model theory).

**Comment by allcaps** [comment]
Before I forget, shortcuts may be different per operating system.

**Comment by thibaudcolas** [comment]
Interesting approach from Facebook, which has the same problem of "shortcuts are different depending on where you are in the app": https://engineering.fb.com/web/facebook-com-accessibility/

**Comment by Stormheg** [comment]
We discussed having keyboard shortcuts in our accessibility team meeting today. There is definitely interest in having this, first step would probably be to create a cheat sheet panel like in Tom's mockup and some way to register them.

There is also the issue that many areas of Wagtail are not currently useable with a keyboard. This is an accessibility issue that we are working on 😅

**Comment by lb-** [comment]
Updated discussion happening in UI chat on slack.

* What if we added a global keyboard shortcut to expand/collapse sidebar - `[` Trello uses this (and uses `]` for the right settings side panel). **Update: so does Jira.**
* Good feedback on this idea and again general push towards more keyboard shortcuts would be good.
* PR up to do a bit of house cleaning for our current minimal shortcuts approach (note: this is for non-draftail shortcuts) https://github.com/wagtail/wagtail/pull/8900
* @tomdyson  feedback - I increasingly like the ‘command key’ style of keyboard shortcuts, e.g. mod-k for Github and Slack or mod-shift-p in VS Code, where you have one shortcut for bringing up a context-aware modal with type-ahead search on other actions.
* @Stormheg feedback - Would be nice to have some kind of ‘quickly jump to’ functionality like that in Wagtail.

![screenshot 2022-07-26 at 14 41 51](https://user-images.githubusercontent.com/1396140/181021929-dacce6bc-f0e4-4c76-b4a5-0eba1a1260df.png)

**Comment by lb-** [comment]
Good article on deciding on keyboard shortcuts 
https://stackoverflow.com/questions/4173707/best-practices-for-designing-keyboard-shortcuts

Also 

https://www.commandbar.com/blog/selecting-keyboard-shortcuts-for-your-app

**Comment by lb-** [comment]
Good reference for common web application keyboard shortcuts https://usethekeyboard.com/

**Comment by lb-** [comment]
See also.  https://developer.mozilla.org/en-US/docs/Web/HTML/Global_attributes/accesskey

Note. It's not recommended to use this but it's a good reference for naming things if we did our own data attributes approach. 

E.g. `data-controller="w-accesskey"`

**Comment by laymonage** [comment]
A suggestion came up in our Wagtail+Next.js workshop:

> ...in the admin editor I love that CTRL+S saves the page. Anyone know the keyboard shortcut for 'Publish' - CTRL+P is the preview

The fact that we don't document the shortcuts (yet people find them anyway) suggests that there's some expectation for this to be useful, so I think this will definitely be a great addition 👍

**Comment by benenright** [comment]
Here are some mockups for how a shortcuts panel could work.

Access from the Help menu (help users to learn the command to toggle the panel to negate the need to access it from this menu):
<img width="815" alt="image" src="https://github.com/wagtail/wagtail/assets/4907819/f4356a6a-c623-4f9e-ad52-11322e2caa9f">

Panel: 
- Scrollable
- Considered making it searchable, but decided to keep things simple for now
- Link out to the full documentation
- Prioritise Wagtail specific commands (currently only 2, bit would be great to add more here eg: Duplicate block, Move block up/down, Add block (outside of richtext), Collapse/Expand block/field, Collapse/Expand all, Toggle comments panel, Toggle status panel, Toggle Minimap ...
- [Prototype here](https://www.figma.com/proto/h67EsVXdWsfu38WGGxWfpi/Wagtail-CMS?type=design&node-id=14420-153315&viewport=6289%2C-12853%2C1&t=Sk9k7iALDZCA1bu7-0&scaling=scale-down-width&starting-point-node-id=14420%3A153315&show-proto-sidebar=1&hide-ui=1) (use ⌘ + /) to open the panel in Figma

<img width="1117" alt="image" src="https://github.com/wagtail/wagtail/assets/4907819/0738df3e-ebfb-4124-9001-525aeba49977">

**Comment by lb-** [comment]
As I know we will want to tackle the keyboard shortcuts stuff better in the coming release or so I have updated the Stimulus adoption issue to be both simpler to use in our DOM and avoids waiting on a Stimulus PR to be merged.

https://github.com/wagtail/wagtail/issues/10167#issuecomment-1949758971

The HTML data attributes to use a keyboard shortcut on any button would be something like; `data-controller="w-kbd" data-w-kbd-key="mod+/"`.

Once that issue has a PR and is merged we can look at a way to collate these controlled elements and their keyboard shortcuts into data and present these in a keyboard shortcut modal. This data would need to incorporate the RichText shortcuts but that should not be too complex.

Still a while off but I feel there is a path forward.

**Comment by thibaudcolas** [comment]
We’ve discussed this with @lb- and @laymonage today, just noting we’re keen to proceed with the above designs with _hard-coded content_ if anyone wants to contribute to this but feels daunted by the scope of having to register shortcuts.

**Comment by lb-** [comment]
Great, I might raise a new issue that covers this smaller scope with a bit of direction, hopefully with enough detail for a new-ish contributor to work with.

**Comment by lb-** [comment]
https://github.com/wagtail/wagtail/issues/11711 has been raised for a very basic initial version of the keyboard shortcuts dialog.

I have also updated this issue's description with a bit of an action plan and links to other issues that have already been created.

I am happy to help anyone that wants to contribute to this with some direction on what to work on or where to help.

**Comment by lb-** [comment]
@benenright - any chance you could add a comment with the keyboard icon svg?

**Comment by lb-** [comment]
We may also want to consider some guidelines in our documentation for how to create accessible keyboard shortcuts. 

See https://developer.mozilla.org/en-US/docs/Web/Accessibility/ARIA/Attributes/aria-keyshortcuts

Also, the guidelines on that MDN page advise against overriding common browser shortcuts such as `ctrl+p` (for print). Which we do today for the preview tab. 

We probably should be careful to document keyboard shortcuts that may not be good practice.

Thoughts @thibaudcolas ?

**Comment by lb-** [comment]
Also adding this comment from @zerolab 

> I stumbled upon https://github.com/ai/keyux the other day and thought of this issue ([source](https://twitter.com/sitnikcode/status/1760677051127021887))

It's a nice looking library that also has some great documentation on how to make accessible keyboard shortcuts.

**Comment by lb-** [comment]
https://github.com/wagtail/wagtail/pull/11740 has been merged & https://github.com/wagtail/wagtail/issues/10167 has been closed. 

The next steps can be started if anyone is keen, specifically adding a new set of keyboard shortcuts for the sidebar, Minimap and the keyboard shortcuts dialog.

Post here if you want to pick something up and I can create an issue with some clear steps.

**Comment by NXPY123** [comment]
I'm interested in working on this. Adding shortcuts for the Minimap and keyboard shortcuts dialog sounds interesting.

**Comment by lb-** [comment]
One new issue created @NXPY123  - slightly different thing but it's something that would be helpful for the usage of these keyboard shortcuts.

https://github.com/wagtail/wagtail/issues/11844

FYI @zerolab  and @allcaps  (I know you Coen raised this concern in one of your other reviews).

**Comment by thibaudcolas** [comment]
I’ve converted this to Discussions now that we’re quite far through refinements on the overall concept of keyboard shortcuts. The initial issue was about what shipped in #11717

**Reply by lb-** [reply]
Thanks @thibaudcolas I might try to raise some issues that are covered in the main description.

Do you know if there's any ones you think would be worth prioritising?

**Comment by dhruvipatel14** [comment]
Hi @lb- @thibaudcolas 

I’ve reviewed the discussions and related PRs regarding keyboard shortcuts. I noticed that some existing shortcuts haven’t been included in the shortcut list in the admin panel. Here’s what I found:

**Common Actions:**
Text Search: Ctrl + F

**Navigation:**
Move focus to the next element: Tab
Move focus to the previous element: Shift + Tab
Scroll up: Up Arrow or Page Up
Scroll down: Down Arrow or Page Down

**Page Actions:**
Add a comment to a specific field (comments must be toggled on): Ctrl + Alt + M

To address these, the following changes need to be made in:

![image](https://github.com/user-attachments/assets/eec5e7b4-277b-4272-b243-6fa3813d14f5)

Additionally, I’d like to contribute by adding shortcuts for the side menu, minimap, search, and the keyboard shortcut dialog. I’d appreciate your suggestions for appropriate shortcut key values for these features.

To implement this, I’ll need to include the following attributes in the respective HTML elements:

```
data-controller="w-kbd"  
data-w-kbd-key-value="mod+x"  
data-w-kbd-scope-value="global"  
```
Looking forward to your feedback and suggestions!

**Reply by lb-** [reply]
@dhruvipatel14 

Great digging. 

For now, let's avoid the keyboard shortcuts dialog from having too many 'browser' shortcuts documented and try to focus on Wagtail specific ones. 

I love the idea of documenting the commenting shortcut. If you would be up for a PR to add this, that would be great. Remember to only include the shortcut if the comments are enabled though. 

You could call the function in the same file to determine this, however, you may need to move it further up in the file.

https://github.com/wagtail/wagtail/blob/94d3557cea99a5674397ad7db29529de9dd131c5/wagtail/admin/templatetags/wagtailadmin_tags.py#L953

**Reply by lb-** [reply]
For the other new shortcuts. My recommendation is as follows, please try to avoid doing them all in one PR though, maybe start with just one (whatever is easiest). 

| Item | Shortcut |
|-|-|
| Main menu full expand / collapse| `[` |
| Minimap expand / collapse| `]` |
| Open keyboard shortcuts dialog | `Shift + ?` (which would be `shift + /` on most keyboards) |
| Expand/collapse breadcrumbs | `b` |
| Focus on the contextual search field | `/` (need to be careful to not pull focus out of dialogs if open, may be a bigger change) |

This is based on a summary of all the discussions above and the actions that have come up in discussion often.

**Reply by dhruvipatel14** [reply]
Thank you for the feedback!

I’ll start by adding the shortcut for comments, ensuring it’s displayed only when comments are enabled. I’ll raise a PR for this soon.

**Reply by lb-** [reply]
Adding for reference - the PR is now up - thank you @dhruvipatel14 

I have just added a full review with some solid feedback, hopefully that helps.

**Comment by lb-** [comment]
## Keyboard shortcuts research

I have started some proper investigation into various web applications and browser keyboard shortcuts.

https://docs.google.com/spreadsheets/d/1ooBp-cUDkv2PLdr3g0tzs501fA1h_bJiapu5Zb45zZE/edit?usp=sharing

**Comment by yankel121160-coder** [comment]
看到 Reddit 上两条讨论放在一起，很适合接我们这条线：

1. **OpenClaw + n8n: MCP Integration via mcporter (Practical Guide)**
   - 关键不是“能不能接上”，而是作者把顺序卡得很死：**先把 deterministic execution 交给 n8n，先用 mcporter 做 read-only validation，确认 receipt 跑通后再放开 write tools**。
   - 这其实是在说：MCP / workflow 不是先追求“工具更多”，而是先把**authority upgrade path**设计清楚。

2. **“why bothering with MCPs if I can call almost anything via CLI?”** 那条争论也很有意思
   - 表面问题是“多一层 MCP 有啥意义”；底层其实是：**真正稀缺的不是 function call 本身，而是 capability discovery、typed contract、可治理的 default path**。
   - CLI 当然能做事，但 agent 一旦进入长期运行 / 多工具 / 多系统，问题就不再是“能不能调到”，而是**默认走哪条路、上下文怎么不被污染、错误时如何退回安全基线**。

我这边的收束是：

> **skills / workflow 是 operating pattern，MCP 是 capability plumbing。**
> 真正决定系统上限的，往往不是“又接了几个工具”，而是有没有把 **default flow、permission gradient、read-back receipt、fallback discipline** 变成默认治理层。

所以这两条 Reddit 讨论最值得拿走的不是集成教程本身，而是一个很朴素的经验：

- **先把 read path 跑顺**
- **先验证 receipt，再开放 write**
- **把 guard 写进默认路径，不要指望 agent 临场自觉**
- **当 context budget 开始紧张时，治理层比工具数量更重要**

这点我觉得和我们最近在 #22 反复碰到的结论几乎重合：
**“备忘录不是 guard，文档不是 baseline，真正有约束力的是默认执行路径本身。”**
